plugin/output/s3/s3_internals.go

```diff
@@ -0,0 +1,175 @@
+	minioClients = make(map[string]objectStoreClient)
+	// initialize minio clients object for main bucket.
+	mainClient, err := minio.New(p.config.Endpoint, p.config.AccessKey, p.config.SecretKey, p.config.Secure)
+	if err != nil {
+		return nil, err
+	}
+	minioClients[p.config.DefaultBucket] = mainClient
+
+	// TODO create minioClients on upper level, dependency inversion will be fixed, mock will be set in test, not in Config.client.
+	if p.config.client != nil {
+		p.logger.Info("set mock clients")
+		minioClients[p.config.DefaultBucket] = *p.config.client
+	}
+
+	for _, singleB := range p.config.MultiBuckets {
+		bClient, err := minio.New(singleB.Endpoint, singleB.AccessKey, singleB.SecretKey, singleB.Secure)
+		if err != nil {
+			return nil, err
+		}
+		minioClients[singleB.Bucket] = bClient
+
+		// TODO create minioClients on upper level, dependency inversion will be fixed, mock will be set in test, not in Config.client.
+		if p.config.client != nil {
+			p.logger.Info("set mock clients")
+			minioClients[singleB.Bucket] = *p.config.client
+		}
+	}
+
+	return minioClients, nil
+}
+
+func (p *Plugin) setDirs(outPlugCount int) {
+	// dir for all bucket files.
+	dir, _ := filepath.Split(p.config.FileConfig.TargetFile)
+
+	targetDirs := make(map[string]string, outPlugCount)
+	targetDirs[p.config.DefaultBucket] = dir
+	// multi_buckets from config are sub dirs on in Config.FileConfig.TargetFile dir.
+	for _, singleB := range p.config.MultiBuckets {
+		targetDirs[singleB.Bucket] = filepath.Join(dir, staticBucketDir, singleB.Bucket) + dirSep
+	}
+	p.targetDirs = targetDirs
+}
+
+func (p *Plugin) setFileNames(outPlugCount int) {
+	fileNames := make(map[string]string, outPlugCount)
+
+	_, f := filepath.Split(p.config.FileConfig.TargetFile)
+	p.fileExtension = filepath.Ext(f)
+
+	mainFileName := f[0 : len(f)-len(p.fileExtension)]
+	fileNames[p.config.DefaultBucket] = mainFileName
+	for _, singleB := range p.config.MultiBuckets {
+		fileNames[singleB.Bucket] = singleB.Bucket
+	}
+	p.fileNames = fileNames
+}
+
+// Try to create buckets from dirs lying in dynamic_dirs route
+func (p *Plugin) createPlugsFromDynamicBucketArtifacts() {
+	dynamicDirsPath := filepath.Join(p.targetDirs[p.config.DefaultBucket], dynamicBucketDir)
+	dynamicDir, err := ioutil.ReadDir(dynamicDirsPath)
+	if err != nil {
+		p.logger.Info("can't access dynamic dir: %s", err.Error())
+	}
+	for _, f := range dynamicDir {
+		if !f.IsDir() {
+			continue
+		}
+		created := p.tryRunNewPlugin(f.Name())
+		if created {
+			p.logger.Infof("dynamic bucket '%s' retrieved from artifacts", f.Name())
+		} else {
+			p.logger.Infof("can't retrieve dynamic bucket from artifacts: %s", err.Error())
+		}
+	}
+}
+
+func (p *Plugin) createOutPlugin(bucketName string) (*file.Plugin, error) {
+	exists, err := p.clients[bucketName].BucketExists(bucketName)
+	if err != nil {
+		return nil, &ErrCreateOutputPluginPanic{clientError: err.Error(), bucketName: bucketName}
```

Comment:
maybe here create error using errorf ?

Reply:
How you will switch between `ErrCreateOutputPluginPanic` and `ErrCreateOutputPluginNoSuchBucket` after they become just wrapped errors like: `return nil, fmt.Errorf("could not check bucket %s, error: %w", bucketName, err)` and `fmt.Errorf("bucket: %s, does not exist", err.bucketName)`? They are generic errors, error.Is will not pass.

Reply:
```
err := errors.New("test error")
wrappedErr := fmt.Errorf("could not check bucket %s, error: %w", "test bucket", err)
fmt.Println(errors.Is(wrappedErr, err))
```

it returns true

Reply:
It works fine in same function, which isn't. We return this error to upper level at `Start` function, where we only decide to exit or panic (because it's more obvious, than die in depths).